lib/src/package_analysis/summary.dart

```diff
@@ -0,0 +1,129 @@
+import 'package:analyzer/dart/analysis/results.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:collection/collection.dart';
+import 'package:path/path.dart' as path;
+
+import 'common.dart';
+import 'normalize_shape.dart';
+import 'shapes.dart';
+
+Future<PackageShape> summarizePackage(
+  PackageAnalysisContext packageAnalysisContext,
+  String packageLocation,
+) async {
+  final package = PackageShape(
+    libraries: <LibraryShape>[],
+    classes: <ClassShape>{},
+  );
+
+  ExecutableShape summarizeExecutableElement(ExecutableElement executableElement) {
+    // ExecutableElement is a superclass of both MethodShape and FunctionShape
+    return ExecutableShape(name: executableElement.name);
+  }
+
+  PropertyShape summarizeProperty(PropertyAccessorElement property) {
+    return PropertyShape(name: property.variable.name);
+  }
+
+  ClassShape summarizeClassElement(ClassElement classElement) {
```

Comment:
Maybe we should start by checking if `package.classes` contains a `ClassShape.id == classElement.id`?

So we don't create duplicates..

Reply:
You're still creating a ClassShape object with the same ID as one that exists in `pacakge.classes` I would suggest not doing that... and just return the value that exists in `package.classes`.

```dart
final shape = package.classes.firstWhere((s) => s.id == classElement.id, else: () => null);
if (shape != null) {
  return shape;
}
```